burn-train/src/metric/dashboard/tui/popup.rs

```diff
@@ -0,0 +1,155 @@
+        }
+    }
+}
+
+/// Popup state.
+pub(crate) enum PopupState {
+    Empty,
+    Full(String, Vec<Callback>),
+}
+
+impl PopupState {
+    /// If the popup is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        matches!(&self, PopupState::Empty)
+    }
+    /// Handle popup events.
+    pub(crate) fn on_event(&mut self, event: &Event) {
+        let mut reset = false;
+
+        match self {
+            PopupState::Empty => {}
+            PopupState::Full(_, callbacks) => {
+                for callback in callbacks.iter() {
+                    if let Event::Key(key) = event {
+                        if let KeyCode::Char(key) = &key.code {
+                            if &callback.trigger == key && callback.callback.call() {
+                                reset = true;
+                            }
+                        }
+                    }
+                }
+            }
+        };
+
+        if reset {
+            *self = Self::Empty;
+        }
+    }
+    /// Create the popup view.
+    pub(crate) fn view(&self) -> Option<PopupView<'_>> {
+        match self {
+            PopupState::Empty => None,
+            PopupState::Full(title, callbacks) => Some(PopupView::new(title, callbacks)),
+        }
+    }
+}
+
+#[derive(new)]
+pub(crate) struct PopupView<'a> {
+    title: &'a String,
+    callbacks: &'a [Callback],
+}
+
+impl<'a> PopupView<'a> {
+    /// Render the view.
+    pub(crate) fn render<'b>(&'a self, frame: &mut TerminalFrame<'b>, size: Rect) {
+        let lines = self
+            .callbacks
+            .iter()
+            .flat_map(|callback| {
+                vec![
+                    Line::from(vec![
+                        Span::from(format!("[{}] ", callback.trigger)).bold(),
+                        Span::from(format!("{} ", callback.title)).yellow().bold(),
+                    ]),
+                    Line::from(Span::from("")),
+                    Line::from(Span::from(callback.description.to_string()).italic()),
+                    Line::from(Span::from("")),
+                ]
+            })
+            .collect::<Vec<_>>();
+
+        let paragraph = Paragraph::new(lines)
+            .alignment(Alignment::Left)
+            .wrap(Wrap { trim: false })
+            .style(Style::default().fg(Color::Gray))
+            .block(
+                Block::default()
+                    .borders(Borders::ALL)
+                    .title_alignment(Alignment::Center)
+                    .style(Style::default().fg(Color::Gray))
+                    .title(Span::styled(
+                        self.title,
+                        Style::default().add_modifier(Modifier::BOLD),
+                    )),
+            );
+
+        let area = centered_rect(60, 40, size);
+        frame.render_widget(paragraph, area);
+    }
+}
+
+// From: https://github.com/ratatui-org/ratatui/blob/main/examples/popup.rs
+fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
```

Comment:
If it's copied, then we should give credit in NOTICES.md file in the root.

Reply:
It's a code snippet adapted from one of their examples, since they are already listed as dependency, I don't think it's necessary to add it into the NOTICE.md file. The link is there for reference if someone needs the example to create popups.

Reply:
Nope. The entire notice is necessary. Just copy and paste the following in out notices file:

https://github.com/ratatui-org/ratatui/blob/main/LICENSE

You can mention the example was adopted.